### ISSUE TITLE
clang silently ignores `trivial_abi` in some cases

### ISSUE THREAD

**Issue by comex** [llvm/llvm-project#39031]
|  |  |
| --- | --- |
| Bugzilla Link | [39683](https://llvm.org/bz39683) |
| Version | trunk |
| OS | All |
| CC | @ahatanak,@dwblaikie,@DougGregor,@hfinkel,@zygoloid |

## Extended Description 
Target: x86_64-apple-darwin18.2.0

In the following code, the class Test is marked `trivial_abi` (and is not a template class), and Clang does not produce any warnings about `trivial_abi` when compiling it, yet for the function `test()` returning a value of that class, Clang's generated code returns it through memory rather than a register.

I'm not sure whether this class should actually be eligible for `trivial_abi` or not, but if not, there definitely should be a warning.

If you change the definition  by adding an additional non-template move constructor, then `trivial_abi` starts being respected:
```diff
+    Test(Test &&other) { x = other.x; }
```
Tested on latest trunk targeting macOS.

```cpp
struct Test {
    int x;
    Test(int x) : x(x) {}
    template <typename T>
    Test(T &&other) { x = other.x; }
    Test(const Test &) = delete;
    template <typename T>
    Test &operator=(T &&other) { x = other.x; }
    Test &operator=(const Test &) = delete;
} __attribute__((trivial_abi));

Test test() { 
    return Test(42);
}
```
